### ISSUE TITLE
[Bug]: sharding works incorrectly when fullyParallel=false

### ISSUE THREAD

**Issue by ayartsev7** [microsoft/playwright#30253]
### Version

1.42.1

### Steps to reproduce
uncompress attached bug_demo.zip, run `npm i` from bug_demo folder
observe 'fullyParallel' property is set to 'false' in playwright.config.js
run
`npx playwright test --shard=1/2`
`npx playwright test --shard=2/2`

### Expected behavior

tests are divided in half

### Actual behavior

all tests are in 1 shard

### Additional context

I'm trying to shard tests between multiple GitLab jobs, and I need tests to be executed sequentially in job.
Inability to shard with `fullyParallel` set to `false` looks like a bug.
[bug_demo.zip](https://github.com/microsoft/playwright/files/14886111/bug_demo.zip)

### Environment

```shell
System:
  OS: Linux 5.10 Ubuntu 20.04.6 LTS (Focal Fossa)
  CPU: (8) x64 Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz
  Memory: 13.62 GB / 15.35 GB
  Container: Yes
Binaries:
  Node: 20.11.0 - /usr/bin/node
  Yarn: 1.22.21 - /usr/bin/yarn
  npm: 10.2.4 - /usr/bin/npm
Languages:
  Bash: 5.0.17 - /usr/bin/bash
npmPackages:
  @playwright/test: ^1.41.2 => 1.41.2 
  playwright: ^1.42.1 => 1.42.1
```

**Comment by dgozman** [comment]
@ayartsev7 Could you please share a repro that we can run locally? Otherwise we won't be able to fix the issue.

**Comment by ayartsev7** [comment]
@dgozman Updated title and description, attached demo repo.

**Comment by dgozman** [comment]
@ayartsev7 This is working as intended. Since shards can run in parallel, Playwright will only shard tests that are designed to run in parallel. By default **test files** run in parallel and therefore shards will contain whole test files. See [parallelism guide](https://playwright.dev/docs/test-parallel) for more details and options to configure parallelism. I'll update the docs to make this connection between shards and parallelism clear.

**Comment by dgozman** [comment]
Closing as per above. If you still encounter problems, please file a new issue by filling in the "Bug Report" template and link to this one.

**Comment by Pinio31** [comment]
Hit this with `fullyParallel: false` (workers reuse shared setup, so we can't enable it): uneven files leave shards empty, example project and 5 of 150 jobs ran 0 tests on one CI run (~6 min wasted).

Opened a draft with an opt-in fix — `PWTEST_SHARD_MODE=balanced` bin-packs whole filesacross shards by test count (deterministic, stateless, default unchanged): https://github.com/microsoft/playwright/pull/41495

Would the team be open to this direction? Happy to iterate.
